File: modules/editor-webview/ios/EditorWebViewPool.swift
import ExpoModulesCore
import os.log
import WebKit

/// One pooled editor page: a live WKWebView that outlives every host it is
/// shown in.
final class EditorWebViewEntry {
  let webView: WKWebView
  let source: String
  var isLoaded = false
  /// The host currently showing the page, if any. Weak: the host is a React
  /// view with its own lifetime; the pool must never keep one alive.
  weak var host: EditorWebViewHostView?

  init(webView: WKWebView, source: String) {
    self.webView = webView
    self.source = source
  }
}

/// Forwards script messages without retaining the pool. The user content
/// controller retains its handlers, so a strong handler → pool → web view →
/// controller chain would be a cycle; react-native-webview's
/// `RNCWeakScriptMessageDelegate` exists for the same reason.
final class ScriptMessageProxy: NSObject, WKScriptMessageHandler {
  private let key: String
  private weak var pool: EditorWebViewPool?

  init(key: String, pool: EditorWebViewPool) {
    self.key = key
    self.pool = pool
  }

  func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
    guard message.name == EditorWebViewPool.messageHandlerName, let body = message.body as? String else {
      return
    }
    pool?.deliver(key, body)
  }
}

/// The pool: WebViews keyed by instance, attached to whichever host is
/// mounted. Everything here runs on the main thread except `postMessage` and
/// `state`, which JS calls synchronously and which only read the table under
/// the lock before hopping to main.
///
/// Messages from a page reach JS as MODULE events (`emit`), never as events on
/// the host view: a host is a React view that comes and goes with every
/// hand-off, and an event dispatched to a view React Native is tearing down at
/// that moment is dropped. The page's `editor-mounted` after a hand-off was
/// lost exactly that way on Android.
final class EditorWebViewPool: NSObject, WKNavigationDelegate {
  static let shared = EditorWebViewPool()
  static let messageHandlerName = "ReactNativeWebView"
  /// The page's origin. Not about:blank: a document with an opaque origin has
  /// every uncaught error masked to "Script error." by WebKit, which turned a
  /// page crash during a hand-off into an empty box with no message anywhere.
  /// Nothing is ever fetched from this host; it exists to give the page an
  /// origin of its own.
  static let pageURL = URL(string: "https://editor.tinycld.invalid/")!
  /// `log stream --predicate 'subsystem == "org.tinycld.editorwebview"'` shows the
  /// pool's lifecycle on a device: creation, attach, load, process death.
  static let log = OSLog(subsystem: "org.tinycld.editorwebview", category: "pool")

  /// Set by the module once it exists; the pool has no module reference of its own.
  var emit: ((String, [String: Any]) -> Void)?

  private var entries: [String: EditorWebViewEntry] = [:]
  private let lock = NSLock()

  private func entry(_ key: String) -> EditorWebViewEntry? {
    lock.lock()
    defer { lock.unlock() }
    return entries[key]
  }

  private func store(_ key: String, _ entry: EditorWebViewEntry?) {
    lock.lock()
    defer { lock.unlock() }
    entries[key] = entry
  }

  private func takeAll() -> [EditorWebViewEntry] {
    lock.lock()
    defer { lock.unlock() }
    let all = Array(entries.values)
    entries.removeAll()
    return all
  }

  private func find(_ webView: WKWebView) -> (String, EditorWebViewEntry)? {
    lock.lock()
    defer { lock.unlock() }
    return entries.first { $0.value.webView === webView }.map { ($0.key, $0.value) }
  }

  // MARK: - Attach / detach (main thread)

  /// Show the page for `key` in `host`, creating and loading it on first use.
  /// Last mount wins: whichever host attaches most recently takes the page.
  func attach(_ key: String, host: EditorWebViewHostView, source: String) {
    let existing = self.entry(key)
    let entry = existing ?? create(key, source: source)
    // The one line that proves a hand-off did not reload: "reused" means the
    // page kept everything it had.
    os_log(.debug, log: Self.log, "attach %{public}@: %{public}@", key, existing == nil ? "created" : "reused")
    if entry.source != source {
      os_log(.error, log: Self.log, "ignoring a different source for %{public}@; the source is fixed at creation", key)
    }
    entry.host = host
    if entry.webView.superview !== host {
      // addSubview moves the view out of its previous superview.
      host.addSubview(entry.webView)
      entry.webView.frame = host.bounds
    }
    host.apply(to: entry.webView)
  }

  /// Drop the host's claim. Guarded on identity so the order in which React
  /// commits a hand-off (old host removed before or after the new one is added)
  /// cannot clear the new host's binding. Deliberately does nothing to the
  /// web view: destroying it here is exactly what react-native-webview does
  /// and what this pool exists to avoid.
  func detach(_ key: String, host: EditorWebViewHostView) {
    guard let entry = entry(key), entry.host === host else {
      return
    }
    entry.host = nil
  }

  func webView(for key: String, ownedBy host: EditorWebViewHostView) -> WKWebView? {
    guard let entry = entry(key), entry.host === host else {
      return nil
    }
    return entry.webView
  }

  // MARK: - Messages

  /// Page → JS. Main thread (WKScriptMessageHandler delivers there).
  func deliver(_ key: String, _ data: String) {
    emit?("onMessage", ["instanceKey": key, "data": data])
  }

  /// JS → page. Any thread; returns whether an instance exists to receive it.
  /// Dispatched on `document`, the one target both platforms use — the page
  /// listens on `window` too, and dispatching on both would deliver twice.
  func postMessage(_ key: String, _ data: String) -> Bool {
    guard let entry = entry(key) else {
      return false
    }
    guard let json = try? JSONSerialization.data(withJSONObject: ["data": data]),
      let literal = String(data: json, encoding: .utf8) else {
      return false
    }
    let script = "document.dispatchEvent(new MessageEvent('message', \(literal)));"
    DispatchQueue.main.async {
      entry.webView.evaluateJavaScript(script, completionHandler: nil)
    }
    return true
  }

  func requestFocus(_ key: String) {
    entry(key)?.webView.becomeFirstResponder()
  }

  func state(_ key: String) -> [String: Any] {
    let entry = self.entry(key)
    return [
      "exists": entry != nil,
      "loaded": entry?.isLoaded ?? false,
      "attached": entry?.host != nil
    ]
  }

  // MARK: - Lifecycle

  func destroy(_ key: String) {
    guard let entry = entry(key) else {
      return
    }
    os_log(.debug, log: Self.log, "destroy %{public}@", key)
    store(key, nil)
    teardown(entry)
  }

  func destroyAll() {
    for entry in takeAll() {
      teardown(entry)
    }
  }

  private func teardown(_ entry: EditorWebViewEntry) {
    let controller = entry.webView.configuration.userContentController
    controller.removeScriptMessageHandler(forName: Self.messageHandlerName)
    controller.removeAllUserScripts()
    entry.webView.navigationDelegate = nil
    entry.webView.removeFromSuperview()
    entry.host = nil
  }

  private func create(_ key: String, source: String) -> EditorWebViewEntry {
    let config = WKWebViewConfiguration()

    let controller = WKUserContentController()
    // The same global react-native-webview installs, so the page is unchanged:
    // it keeps calling window.ReactNativeWebView.postMessage(String).
    let shim = WKUserScript(
      source: """
      window.\(Self.messageHandlerName) = window.\(Self.messageHandlerName) || {};
      window.\(Self.messageHandlerName).postMessage = function (data) {
        window.webkit.messageHandlers.\(Self.messageHandlerName).postMessage(String(data));
      };
      """,
      injectionTime: .atDocumentStart,
      forMainFrameOnly: true
    )
    controller.addUserScript(shim)
    controller.add(ScriptMessageProxy(key: key, pool: self), name: Self.messageHandlerName)
    config.userContentController = controller
    config.allowsInlineMediaPlayback = true
    // An editor must not turn typed phone numbers and addresses into links.
    config.dataDetectorTypes = []

    let webView = WKWebView(frame: .zero, configuration: config)
    webView.navigationDelegate = self
    webView.allowsLinkPreview = false
    webView.allowsBackForwardNavigationGestures = false
    webView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    webView.scrollView.contentInsetAdjustmentBehavior = .never
    EditorWebViewKeyboardHacks.hideInputAccessoryView(webView)
    EditorWebViewKeyboardHacks.disableKeyboardDisplayRequiresUserAction(webView)

    let entry = EditorWebViewEntry(webView: webView, source: source)
    store(key, entry)
    webView.loadHTMLString(source, baseURL: Self.pageURL)
    return entry
  }

  // MARK: - WKNavigationDelegate

  func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
    guard let found = find(webView) else {
      return
    }
    let (key, entry) = found
    entry.isLoaded = true
    os_log(.debug, log: Self.log, "loaded %{public}@", key)
    emit?("onLoad", ["instanceKey": key])
  }

  /// The content process died (memory pressure, a WebKit crash). Reload the
  /// same source; the page posts `editor-ready` again as it boots and the JS
  /// host re-sends init for the new page epoch.
  func webViewWebContentProcessDidTerminate(_ webView: WKWebView) {
    guard let found = find(webView) else {
      return
    }
    let (key, entry) = found
    entry.isLoaded = false
    os_log(.error, log: Self.log, "content process terminated %{public}@; reloading", key)
    webView.loadHTMLString(entry.source, baseURL: Self.pageURL)
    emit?("onProcessGone", ["instanceKey": key])
  }

  func webView(
    _ webView: WKWebView,
    decidePolicyFor navigationAction: WKNavigationAction,
    decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
  ) {
    // The page is the whole product; a link inside it must not navigate the
    // editor away. Only the initial (and any reload) load of the source passes.
    let isOwnLoad = navigationAction.request.url == Self.pageURL
    decisionHandler(isOwnLoad ? .allow : .cancel)
  }
}

File: modules/editor-webview/android/src/main/java/org/tinycld/editorwebview/EditorWebViewPool.kt
package org.tinycld.editorwebview

import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.core.os.bundleOf
import org.json.JSONObject

private const val TAG = "EditorWebView"

/**
 * The pool: WebViews keyed by instance, attached to whichever host is mounted.
 * Everything here runs on the main thread except [postMessage] and [state],
 * which JS calls synchronously and which only read the table under the lock
 * before posting to main.
 *
 * Messages from a page reach JS as MODULE events ([emit]), never as events on
 * the host view: a host is a React view that comes and goes with every
 * hand-off, and an event dispatched to a view React Native is tearing down at
 * that moment is dropped. The page's `editor-mounted` after a hand-off was lost
 * exactly that way.
 */
object EditorWebViewPool {
  /** The page's origin — see the iOS pool for why it is not about:blank. */
  const val PAGE_URL = "https://editor.tinycld.invalid/"

  /** One pooled editor page: a live WebView that outlives every host it is shown in. */
  class Entry(var webView: PooledWebView, val source: String) {
    var isLoaded = false
    /** The host currently showing the page, if any. */
    var host: EditorWebViewHost? = null
  }

  /** Set by the module once it exists; the pool has no module reference of its own. */
  var emit: ((String, Bundle) -> Unit)? = null

  private val entries = HashMap<String, Entry>()
  private val mainHandler = Handler(Looper.getMainLooper())

  @Synchronized
  private fun entry(key: String): Entry? = entries[key]

  @Synchronized
  private fun store(key: String, entry: Entry?) {
    if (entry == null) entries.remove(key) else entries[key] = entry
  }

  @Synchronized
  private fun takeAll(): List<Entry> {
    val all = entries.values.toList()
    entries.clear()
    return all
  }

  @Synchronized
  private fun find(webView: PooledWebView): Pair<String, Entry>? =
    entries.entries.firstOrNull { it.value.webView === webView }?.let { it.key to it.value }

  // region Attach / detach (main thread)

  /**
   * Show the page for [key] in [host], creating and loading it on first use.
   * Last mount wins: whichever host attaches most recently takes the page.
   */
  fun attach(key: String, host: EditorWebViewHost, source: String) {
    val existing = entry(key)
    val entry = existing ?: create(key, webViewContext(host), source)
    // The one line that proves a hand-off did not reload: "reused" means the
    // page kept everything it had. `adb logcat -s EditorWebView`.
    Log.d(TAG, "attach $key: ${if (existing == null) "created" else "reused"}")
    if (entry.source != source) {
      Log.w(TAG, "ignoring a different source for instance $key; the source is fixed at creation")
    }
    entry.host = host
    val webView = entry.webView
    if (webView.parent !== host) {
      // Leaving a view group clears focus. A focus asked for during the
      // hand-off (the surface asks the moment it acquires) can land before the
      // move, so carry it across; without this the caret is in the page but
      // keystrokes go nowhere.
      val hadFocus = webView.isFocused
      (webView.parent as? ViewGroup)?.removeView(webView)
      host.addView(
        webView,
        ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
      )
      // React Native ignores the requestLayout a late addView triggers, so the
      // child would stay 0×0 until the next RN layout pass.
      host.measureAndLayout()
      if (hadFocus) requestFocus(key)
    }
    host.apply(webView)
  }

  /**
   * Drop the host's claim. Guarded on identity so the order in which React
   * commits a hand-off cannot clear the new host's binding. Removes the WebView
   * from a dropped host's view group but never destroys it — destroying is
   * exactly what react-native-webview does and what this pool exists to avoid.
   */
  fun detach(key: String, host: EditorWebViewHost) {
    val entry = entry(key) ?: return
    if (entry.host !== host) return
    entry.host = null
    if (entry.webView.parent === host) {
      host.removeView(entry.webView)
    }
  }

  // endregion

  // region Messages

  /** Page → JS. Main thread. */
  fun deliver(key: String, data: String) {
    if (entry(key) == null) return
    emit?.invoke("onMessage", bundleOf("instanceKey" to key, "data" to data))
  }

  /**
   * JS → page. Any thread; returns whether an instance exists to receive it.
   * The dispatch script is react-native-webview's, which already targets
   * `document` — the one target both platforms use.
   */
  fun postMessage(key: String, data: String): Boolean {
    val entry = entry(key) ?: return false
    val eventInit = JSONObject().put("data", data).toString()
    val script = "(function () {" +
      "var data = " + eventInit + ";" +
      "var event = new MessageEvent('message', data);" +
      "document.dispatchEvent(event);" +
      "})();"
    mainHandler.post { entry.webView.evaluateJavascript(script, null) }
    return true
  }

  fun requestFocus(key: String) {
    val webView = entry(key)?.webView ?: return
    webView.requestFocus()
    // react-native-webview stops at requestFocus; after a re-parent that alone
    // does not reliably bring the keyboard up.
    val imm = webView.context.getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
    imm?.showSoftInput(webView, InputMethodManager.SHOW_IMPLICIT)
  }

  fun state(key: String): Map<String, Any> {
    val entry = entry(key)
    return mapOf(
      "exists" to (entry != null),
      "loaded" to (entry?.isLoaded ?: false),
      "attached" to (entry?.host != null)
    )
  }

  // endregion

  // region Lifecycle

  fun destroy(key: String) {
    val entry = entry(key) ?: return
    Log.d(TAG, "destroy $key")
    store(key, null)
    teardown(entry)
  }

  fun destroyAll() {
    takeAll().forEach { teardown(it) }
  }

  private fun teardown(entry: Entry) {
    entry.host = null
    (entry.webView.parent as? ViewGroup)?.removeView(entry.webView)
    entry.webView.removeJavascriptInterface(PooledWebView.BRIDGE_NAME)
    entry.webView.destroy()
  }

  /**
   * The Activity when there is one: it is what gives the WebView its IME and
   * text-selection action mode, the same context react-native-webview's
   * themed context resolves to.
   */
  private fun webViewContext(host: EditorWebViewHost): Context =
    host.appContext.currentActivity ?: host.context

  private fun create(key: String, context: Context, source: String): Entry {
    val webView = PooledWebView(context, key)
    val entry = Entry(webView, source)
    store(key, entry)
    webView.loadDataWithBaseURL(PAGE_URL, source, "text/html", "utf-8", null)
    return entry
  }

  /** From the WebView client, main thread. */
  fun onPageFinished(webView: PooledWebView) {
    val (key, entry) = find(webView) ?: return
    entry.isLoaded = true
    Log.d(TAG, "loaded $key")
    emit?.invoke("onLoad", bundleOf("instanceKey" to key))
  }

  /**
   * The render process died. By contract the WebView instance is unusable
   * afterwards, so unlike iOS this is a rebuild: a fresh WebView with the same
   * source takes the dead one's place in the same host. The page posts
   * `editor-ready` again as it boots and the JS host re-sends init for the new
   * page epoch. Returns true because returning false would kill the app.
   */
  fun onRenderProcessGone(webView: PooledWebView): Boolean {
    val (key, entry) = find(webView) ?: return true
    Log.e(TAG, "render process gone for $key; rebuilding")
    val host = entry.host
    (webView.parent as? ViewGroup)?.removeView(webView)
    webView.removeJavascriptInterface(PooledWebView.BRIDGE_NAME)
    webView.destroy()

    val replacement = PooledWebView(webView.context, key)
    entry.webView = replacement
    entry.isLoaded = false
    replacement.loadDataWithBaseURL(PAGE_URL, entry.source, "text/html", "utf-8", null)
    if (host != null) {
      host.addView(
        replacement,
        ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
      )
      host.measureAndLayout()
      host.apply(replacement)
    }
    emit?.invoke("onProcessGone", bundleOf("instanceKey" to key))
    return true
  }

  // endregion
}
